Span Finder Suggester

thomashacker opened the pull request:
This `PR` adds a new experimental component for learning span boundaries and a custom suggester function for spancat.
It further adds a spaCy project showcasing how to use the SpanFinder component on 3 different datasets (Healthsea, ToxicSpans, Genia) with 2 configurations (tok2vec & transformer). The project also provides the possibility to train spancat with ngram and compare it to SpanFinder with a custom evaluation script that calculates the performance and overall coverage of the suggester functions.

## Features
- spaCy project for comparing SpanFinder vs Ngram
- SpanFinder model
- SpanFinder component
- SpanFinder suggester
- Unit tests for component, model and suggester

adrianeboyd commented on spacy_experimental/span_boundary_detection/sbd_component.py:
```diff
@@ -0,0 +1,301 @@
+
+
+@Language.factory(
+    "spacy-experimental_span_boundary_detection_component_v1",
```
May I suggest: `experimental_span_boundary_detector`?

The factory names usually don't include `spacy` or version numbers.

adrianeboyd commented:
Very nice! This is going to be annoying, but I think that we need to avoid the abbreviation `sbd` because it's used too much for "sentence boundary detection".

kadarakos commented on spacy_experimental/span_boundary_detection/sbd_component.py:
```diff
@@ -0,0 +1,301 @@
+        self.model = model
+        self.name = name
+        self.scorer = scorer
+        Token.set_extension("span_start", default=0, force=True)
```
Currently for the `iob` tags the `Token` has attributes `Token.ent_iob` and `Token.ent_iob_` https://spacy.io/api/token#attributes. It seems to me that each `Token` could be either `B`, `E` or nothing so maybe `Token.span_boe`? Not sure

thomashacker replied:
So you mean adding a new attribute to the Token? `Token.span_boe` 😄

kadarakos replied:
yeah something like that, rather than having a `span_start` and a `span_end` it would be more like the way the `Token.ent_iob` works.

adrianeboyd replied:
There are a lot of drawbacks to representing span annotation as token attributes I'd prefer not to add any more of it, and also drawbacks to how integers are used internally for `ent_iob` in particular.

What are the exact requirements here?

thomashacker replied:
I'll rebuild the whole architecture so that the SpanFinder assigns a SpanGroup with all predicted spans and the SpanFinderSuggester uses the SpanGroup to create the Ragged object for the spancat prediction. This way we don't need another token attribute 😄

kadarakos commented on spacy_experimental/span_boundary_detection/sbd_component.py:
```diff
@@ -0,0 +1,301 @@
+            offset += length
+
+        for doc, score_doc in zip(docs, scores_per_doc):
+            for token, score_token in zip(doc, score_doc):
```
maybe this variable could be `doc_scores` and the inner variable `token_score`. Then the

```
token_score[0] > self.cfg["threshold"]
```

"reads" nicer perhaps.

kadarakos commented on spacy_experimental/span_boundary_detection/sbd_component.py:
```diff
@@ -0,0 +1,301 @@
+                else:
+                    score_token[1] = 0
+
+                token._.span_start = score_token[0]
```
If neither of these are true then `token.span_boe` would be `0` and otherwise `1` or `2`. I don't think its crucial, maybe later for consistency.

kadarakos commented on spacy_experimental/span_boundary_detection/sbd_component.py:
```diff
@@ -0,0 +1,301 @@
+        losses.setdefault(self.name, 0.0)
+        references = [eg.reference for eg in examples]
+        set_dropout_rate(self.model, drop)
+        scores, backprop_scores = self.model.begin_update(references)
```
I think all other components do the update on the `eg.predicted` https://github.com/explosion/spaCy/blob/master/spacy/pipeline/spancat.py#L339, but I'm not exa tly sure why.

thomashacker replied:
It makes sense because `example.predicted` are the docs that don't have the annotations, but since we're only feeding in the raw docs to model it makes no difference for the predictions. I'm using the `example.reference` because I also need them to calculate the score/loss, so I only have to do `references = [eg.reference for eg in examples]` once. But I'm interested in what the other's think? It would certainly look better if I would use `example.predicted`

kadarakos commented on spacy_experimental/span_boundary_detection/sbd_model.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def flattener() -> Model[Floats1d, Floats1d]:
+    def forward(model: Model[Floats1d, Floats1d], X, is_train: bool) -> Floats1d:
```
What is the type of `X`? `Floats2d`?

kadarakos commented on spacy_experimental/span_boundary_detection/sbd_suggester.py:
```diff
@@ -0,0 +1,54 @@
+                for end in ends:
+                    if start < end and (start, end) not in cache:
+                        spans.append([start, end])
+                        cache[(start, end)] = True
```
the `cache` here could just be a `set()` and then `if (start, end) not in cache` still makes sense.

thomashacker replied:
That's so embarrassing, how did I not know of `sets` before? I'm using this workaround with `dicts` all the TIME! This is super useful 😆

thomashacker commented:
@adrianeboyd how about instead of "sbd" -> "span_bd"? 😄

svlandeg reviewed (commented):
This is great Edi!

From a practical point, I'm wondering whether we need all the different config files in all the different subdirectories? It's great to support various different datasets, but what if we could make the decision very early on:
- You process one of the three available datasets, with one of the three specific commands (cf below)
- The processed `.spacy` output is stored in the same subdir of the project, regardless of what dataset had been processed
- Now, all consequent steps don't have to bother about where the data came from originally

Is that feasible? Or are there things you're doing different in the scripts, depending on what dataset it was (besides preprocessing)

thomashacker commented on projects/span_boundary_detection/README.md:
```diff
@@ -0,0 +1,62 @@
+
+| Workflow | Steps |
+| --- | --- |
+| `preprocess` | `preprocess_healthsea` &rarr; `preprocess_toxic` &rarr; `preprocess_genia` |
+| `analyze` | `analyze_healthsea` &rarr; `analyze_toxic` &rarr; `analyze_genia` |
```
Yeah initially I only used one config for all the datasets but then quickly realized that the configs need to be adjusted to the dataset (e.g. ngram range)... 

I thought it would be more user-friendly if they could just change the dataset variable to the dataset that they want to train on

thomashacker replied:
I think the only important difference is the ngram size... But I think you're right. I'll just use one config for all datasets to make it more user-friendly! 😄

thomashacker commented:
Alright, I've renamed every reference to `SpanFinder` 😄 I've rebuilt the `set_annotation()` logic so that the `SpanFinder` component directly produces spans and saves them to `doc.spans["span_finder_candidates"]` instead of having an additional token attribute. The suggester uses `doc.spans["span_finder_candidates"]` to produce the `Ragged` for the `spancat`.

I've also adjusted the spaCy project, added more commands, and reduced the number of configs, so that now there is only one config for all datasets.

adrianeboyd commented:
As the next minor adjustment, I think that `span_finder_candidates` should be a configurable setting.

svlandeg commented on projects/span_boundary_detection/project.yml:
```diff
@@ -0,0 +1,187 @@
+      - "data/toxic_${vars.train}.spacy"
+      - "data/toxic_${vars.dev}.spacy"
+
+  # Required to have spacy-span-analyzer installed (https://github.com/ljvmiranda921/spacy-span-analyzer)
```
cf https://github.com/explosion/spaCy/pull/10668

svlandeg commented on projects/span_finder/project.yml:
```diff
@@ -0,0 +1,171 @@
+# run them via "spacy project run [workflow]". If a commands's inputs/outputs
+# haven't changed, it won't be re-run.
+workflows:
+  healthsea_data:
+    - preprocess_healthsea
+    - analyze_healthsea
+  toxic_data:
+    - preprocess_healthsea
+    - analyze_healthsea
+  genia_data:
+    - preprocess_healthsea
+    - analyze_healthsea
```
copy-paste error ;-)

svlandeg commented on projects/span_finder/project.yml:
```diff
@@ -0,0 +1,171 @@
+vars:
+  config: "config_tok2vec" # config_tok2vec, config_trf
+  dataset: "healthsea" # healthsea, toxic, genia
+  suggester: "ngram" # ngram, span_finder
```
Considering this project is to showcase the `span_finder` functionality, maybe that should be the default value? :-)

svlandeg reviewed (commented):
I've been running the project - runs smoothly! Had just a few more small comments. 

(I'll continue with some more reviews later today)

thomashacker commented on projects/span_finder/project.yml:
```diff
@@ -0,0 +1,171 @@
+  - name: "evaluate_spancat"
+    help: "Evaluate a trained spancat model"
+    script:
+      - "python -m spacy evaluate ${vars.spancat_model} data/${vars.dataset}_${vars.dev}.spacy --output metrics/spancat_${vars.dataset}_${vars.config}_${vars.suggester}.json --gpu-id ${vars.gpu_id}"
+      - "python scripts/evaluation.py ${vars.span_key} ${vars.spancat_model} data/${vars.dataset}_${vars.dev}.spacy"
```
I've removed the metrics from the custom script :smile:

svlandeg commented on projects/span_finder/README.md:
```diff
@@ -0,0 +1,69 @@
+| `suggester`  | Choose between two suggester architectures (SpanFinder, Ngram)                              |
+| `train`      | Choose a filename for your training data                                                    |
+| `dev`        | Choose a filename for your development data                                                 |
+| `span_key`   | Choose a key to specify the SpanGropup for the spancat component to save the predictions to |
```
```suggestion
| `span_key`   | Choose a key to specify the SpanGroup for the spancat component to save the predictions to  |
```

svlandeg commented on projects/span_finder/configs/ngram/config_tok2vec.cfg:
```diff
@@ -0,0 +1,140 @@
+
+[nlp]
+lang = "en"
+pipeline = ["spancat"]
```
Out of curiosity, why did you use a builtin `tok2vec` here instead of a separate component like you did for the `span_finder` config?

thomashacker replied:
I find the config to be more "cleaner" this way 😇

svlandeg commented on setup.cfg:
```diff
@@ -19,17 +19,22 @@ install_requires =
 spacy_tokenizers =
     spacy-experimental.char_pretokenizer.v1 = spacy_experimental.char_tokenizer.char_pretokenizer:char_pretokenizer_v1
+spacy_misc =
+    experimental.span_finder_suggester.v1 = spacy_experimental.span_finder.span_finder_suggester:build_span_finder_suggester
```
For consistency, I'd prefix all of these new ones with `spacy-experimental` like we've done for the current factories in this package.

thomashacker commented on projects/span_finder/scripts/preprocessing/preprocess_genia.py:
```diff
@@ -0,0 +1,81 @@
+    for docs in zip(*docs_per_level):
+        spans = [ent for doc in docs for ent in doc.ents]
+        doc = docs[0]
+        group = SpanGroup(doc, name="sc", spans=spans)
```
@ljvmiranda921 could you help us out here 😇🙏

ljvmiranda921 commented on projects/span_finder/scripts/preprocessing/preprocess_genia.py:
```diff
@@ -0,0 +1,80 @@
+
+    for docs in zip(*docs_per_level):
+        spans = [ent for doc in docs for ent in doc.ents]
+        doc = docs[0]
```
Hi @svlandeg @thomashacker (I recall seeing a notif that asks about this line of code, but I can't seem to find it now so I'll add it here). So the way the GENIA IOB was originally constructed is that it has 4 "levels" due to their overlapping nature. What I did was create a `Doc` with the entities per each level (using `conll_ner_to_docs`)

However, we want all entities in a single `Doc`, so the way I "merged" them is to put all the entities into the `.spans` of the first `Doc` (arbitrary choice).

adrianeboyd commented on setup.cfg:
```diff
@@ -19,17 +19,22 @@ install_requires =
 spacy_architectures =
     spacy-experimental.Bilinear.v1 = spacy_experimental.biaffine_parser.bilinear:build_bilinear
     spacy-experimental.PairwiseBilinear.v1 = spacy_experimental.biaffine_parser.pairwise_bilinear:build_pairwise_bilinear
+    spacy-experimental.span_finder_model.v1 = spacy_experimental.span_finder.span_finder_model:build_finder_model
```
The architectures usually have names like `spacy.TextCatEnsemble.v2` or `spacy.Tagger.v1` and don't include `model `in the name, so something like `spacy-experimental.SpanFinder.v1`.

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,298 @@
+            for start in starts:
+                for end in ends:
+                    if start < end:
+                        doc.spans[self.candidates_key].append(doc[start:end])
```
Doesn't this really explode for longer docs?

thomashacker replied:
You're right. I'll try to think of a way to make this more efficient!

thomashacker replied:
I've rewritten the code a bit and it's slightly faster now, I couldn't think of a better way yet.

svlandeg commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,298 @@
+    )
+
+
+@registry.scorers("spacy-experimental.span_finder_scorer.v1")
```
Because we have them as entry points in `setup.cfg`, we can remove the decorators from these functions (also to ensure that the names don't go out-of-sync)

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,295 @@
+            scores_per_doc.append(scores[offset : offset + length])
+            offset += length
+
+        for doc, doc_scores in zip(docs, scores_per_doc):
+            doc.spans[self.candidates_key] = []
+            for start_token, start_token_score in zip(doc, doc_scores):
+                if start_token_score[0] > self.cfg["threshold"]:
+                    for end_token, end_token_score in zip(doc, doc_scores):
+                        if (
+                            end_token_score[1] > self.cfg["threshold"]
+                            and end_token.i + 1 > start_token.i
+                        ):
+                            doc.spans[self.candidates_key].append(
+                                doc[start_token.i : end_token.i + 1]
+                            )
```
Although it's still a concern, my concern is less the efficiency of this particular loop than the efficiency of potentially making a very large number of very long suggestions that you probably actually don't want that you're passing to `spancat`. I think it might make sense to add a `max_span_length` parameter or something along those lines?

Other options I could imagine that are a little more complicated:

* not crossing sentence boundaries
* use the N (say, 3) nearest ends for a particular start

Obviously this depends on your use case, but I think the current approach is only suitable for very short texts.

thomashacker replied:
I've added two new parameters to the `SpanFinder`  and added more information to the README in [6422926](https://github.com/explosion/spacy-experimental/pull/10/commits/6422926e6af418467de92d2595e7d99eccd29f99) 
- `max_length`
- `min_length`
> When set to 0, the limitation is ignored.

adrianeboyd replied:
I would go back to a version where `starts` and `ends` were extracted initially to avoid repeatedly checking thresholds.

I think it would be faster to avoid having an extra function call to check the length constraints and you can build it shortcuts a bit more easily.

Try to order the checks so that very fast checks are run first and shortcut as much as possible:

- check whether end <= start
- check whether the limits are set
- once you have seen one end that's beyond the max_length away from the current start, you can shortcut the rest of that loop

thomashacker replied:
Thanks! I think, I was able to shorten everything quite a bit, I'm surprised that I didn't think of it earlier. Updated in [c424769](https://github.com/explosion/spacy-experimental/pull/10/commits/c4247698000669895c834bdd0f1c7a212150b7a3)

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,323 @@
+            for start in starts:
+                for end in ends:
+                    if start < end + 1:
+                        if (
+                            self.cfg["min_length"] <= 0
+                            or start >= self.cfg["min_length"]
+                        ):
+                            if (
+                                self.cfg["max_length"] <= 0
+                                or end <= self.cfg["max_length"]
+                            ):
+                                doc.spans[self.cfg["candidates_key"]].append(
+                                    doc[start : end + 1]
+                                )
```
If you have two simple nested `if`s you can just use one `if` with `and` instead.

Right now this is comparing the `start`/`end` indices rather than the span length.

Remember to break out of the loop as soon as you see a span length that's longer than `max_length`.

thomashacker replied:
You're absolutely right! I don't know why it's taking me so long for this 😃 Updated in [ad9736e](https://github.com/explosion/spacy-experimental/pull/10/commits/ad9736efd0c998ba98e22abe1a8a63d8c08c783d)

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,323 @@
+
+            for start in starts:
+                for end in ends:
+                    if start < end + 1:
+                        span_length = end + 1 - start
+                        if (
+                            self.cfg["min_length"] <= 0
+                            or span_length >= self.cfg["min_length"]
+                        ) and (
+                            self.cfg["max_length"] <= 0
+                            or span_length <= self.cfg["max_length"]
+                        ):
+                            doc.spans[self.cfg["candidates_key"]].append(
+                                doc[start : end + 1]
+                            )
```
```suggestion
                    span_length = end + 1 - start
                    if span_length > 0:
                        if (
                            self.cfg["min_length"] <= 0
                            or span_length >= self.cfg["min_length"]
                        ) and (
                            self.cfg["max_length"] <= 0
                            or span_length <= self.cfg["max_length"]
                        ):
                            doc.spans[self.cfg["candidates_key"]].append(
                                doc[start : end + 1]
                            )
                        elif (
                            self.cfg["max_length"] > 0
                            and span_length > self.cfg["max_length"]
                        ):
                            break
```

adrianeboyd replied:
Maybe you can add a unit test where you test just `set_annotations` with mock scores?

adrianeboyd commented on spacy_experimental/span_finder/tests/test_span_finder.py:
```diff
@@ -68,6 +70,40 @@ def test_span_finder_component_span_lengths():
 
+def test_span_finder_component_set_annotations_span_lengths():
+
+    test_min_length = 2
+    test_max_length = 3
```
Can you parameterize this with checks for each as 0, each set, both set and also include an explicit check for the correct number of spans in each case (also checking the min/max for good measure).

adrianeboyd commented on spacy_experimental/span_finder/tests/test_span_finder.py:
```diff
@@ -68,6 +70,40 @@ def test_span_finder_component_span_lengths():
+    span_finder.set_annotations([doc], scores)
+
+    assert doc.spans[DEFAULT_CANDIDATES_KEY]
+    assert len(
+        doc.spans[DEFAULT_CANDIDATES_KEY][0]
+    ) >= test_min_length and test_max_length >= len(
+        doc.spans[DEFAULT_CANDIDATES_KEY][0]
+    )
```
This ordering in the comparisons threw me a bit. It won't let me suggest this directly (not sure why), so something more like:

```python
assert all(test_min_length <= len(span) <= test_max_length for span in doc.spans[DEFAULT_CANDIDATES_KEY])
```

adrianeboyd commented on README.md:
```diff
@@ -216,6 +216,102 @@ annotating_components = ["senter"]
+
+The SpanFinder is a new experimental component that tries to learn span boundaries by tagging potential start and end tokens. It's an ML approach on suggester functions to produce fewer and more precise span candidates. 
+
+`SpanFinder` requires three parameters for initialization:
+- `threshold`: Threshold of the required probabilty for predicted spans.
+- `candidates_key`: Name of the [SpanGroup](https://spacy.io/api/spangroup) in which the predicted spans are saved to.
+- `max_length`: Max length of the predicted spans (no limit when set to 0).
+- `min_length`: Min length of the predicted spans (no limit when set to 0).
```
```suggestion
`SpanFinder` uses the following parameters:
- `threshold`: Probability threshold for predicted spans.
- `candidates_key`: Name of the [SpanGroup](https://spacy.io/api/spangroup) the predicted spans are saved to.
- `max_length`: Max length of the predicted spans. No limit when set to `0`. Defaults to `0`.
- `min_length`: Min length of the predicted spans. No limit when set to `0`. Defaults to `0`.
```

adrianeboyd commented on README.md:
````diff
@@ -216,6 +216,102 @@ annotating_components = ["senter"]
+candidates_key = ${components.span_finder.candidates_key}
+```
+
+This package includes a [spaCy project](./projects/span_finder) which shows how to train and use the `SpanFinder` together with `Spancat`.
````
```suggestion
This package includes a [spaCy project](./projects/span_finder) which shows how to train and use the `SpanFinder` together with `SpanCategorizer`.
```

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,327 @@
+            [doc.predicted for doc in examples], candidates_key
+        )
+
+        scorer = PRFScore()
+
+        for prediction_doc, reference_doc in zip(predictions, references):
+            for prediction, reference in zip(prediction_doc, reference_doc):
+                if prediction in reference_doc:
+                    scorer.tp += 1
+                else:
+                    scorer.fp += 1
+
+                if reference not in prediction_doc:
+                    scorer.fn += 1
+
+        # Assemble final result
+        final_scores: Dict[str, Any] = {
+            f"span_finder_f": scorer.fscore,
+            f"span_finder_r": scorer.precision,
+            f"span_finder_p": scorer.recall,
+        }
```
Hmm, I don't think this needs a custom scorer? This seems like it should be extremely similar to `spancat_score` just unlabeled?

adrianeboyd replied:
After looking a bit more, I understand a bit more why you implemented it this way, but it's not going to work in `get_span_references` or in `_get_reference` to use all the spans in the reference document rather than specifying a particular spans key to learn from. Let me think a bit more about how this could work...

adrianeboyd commented on projects/span_finder/project.yml:
```diff
@@ -0,0 +1,136 @@
+# Assets that should be downloaded or available in the directory. We're shipping
+# them with the project, so they won't have to be downloaded. But the
+# 'project assets' command still lets you verify that all required files are available.
+assets:
+  - dest: "assets/healthsea_ner.jsonl"
+    description: "Annotations from the Healthsea dataset"
+    url: https://github.com/explosion/healthsea/blob/main/project/assets/ner/annotation.jsonl
+
+  - dest: "assets/toxic_spans_annotations.csv"
+    description: "Annotations from the ToxicSpans dataset"
+    url: https://github.com/ipavlopoulos/toxic_spans/tree/master/data/annotations.csv
+
+  - dest: "assets/toxic_spans.csv"
+    description: "Spans from the ToxicSpans dataset"
+    url: https://github.com/ipavlopoulos/toxic_spans/tree/master/data/spans.csv
+
+  - dest: "assets/toxic_spans_comments.csv"
+    description: "Comments from the ToxicSpans dataset"
+    url: https://github.com/ipavlopoulos/toxic_spans/tree/master/data/comments.csv
+
+  - dest: "assets/genia_train.iob"
+    description: "Training annotations from the Genia dataset"
+    url: "https://github.com/thecharm/boundary-aware-nested-ner/tree/master/Our_boundary-aware_model/data/genia/genia.train.iob2"
+
+  - dest: "assets/genia_dev.iob"
+    description: "Dev annotations from the Genia dataset"
+    url: "https://github.com/thecharm/boundary-aware-nested-ner/tree/master/Our_boundary-aware_model/data/genia/genia.dev.iob2"
```
I think the URLs should be converted to the raw URLs.

adrianeboyd commented:
So there was a problem in both training and scoring where it was using every single span in `doc.spans` as a reference span rather than the spans from a particular key. I added an option called `reference_key` as an initial way to handle this, but it might not be the only solution, or we might some further renaming.

As a developer, I want to call the options `predicted_key` and `reference_key`, but I don't think that will be clear for users, who might find something like `candidates_key` and `training_key` easier to understand.

I also want to refactor the options so they're not saved in `self.cfg` so that there aren't multiple sources of settings in the serialized version and updating the config will update any of these settings, e.g. for a user who want so save the candidates to a different key.

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,322 @@
+    **kwargs,
+) -> Dict[str, Any]:
+    kwargs = dict(kwargs)
+    attr_prefix = "span_finder_"
```
I think this still isn't parameterized correctly.

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,322 @@
+        losses[self.name] += loss
+        return losses
+
+    def get_loss(self, examples, scores) -> Tuple[float, float]:
+        """Find the loss and gradient of loss for the batch of documents and
+        their predicted scores.
+        examples (Iterable[Examples]): The batch of examples.
+        scores: Scores representing the model's predictions.
+        RETURNS (Tuple[float, float]): The loss and the gradient.
+        """
+        references = [eg.reference for eg in examples]
+        reference_results = self.model.ops.asarray(
+            self._get_reference(references), dtype=float32
+        )
+        d_scores = scores - reference_results
+        loss = float((d_scores**2).sum())
+        return loss, d_scores
+
+    def _get_reference(self, docs) -> Floats2d:
+        """Create a reference list of token probabilities for calculating loss"""
+        reference_results = []
+        for doc in docs:
+            start_indices = set()
+            end_indices = set()
+
+            if self.reference_key in doc.spans:
+                for span in doc.spans[self.reference_key]:
+                    start_indices.add(span.start)
+                    end_indices.add(span.end - 1)
+
+            for token in doc:
+                reference_results.append(
+                    (
+                        1 if token.i in start_indices else 0,
+                        1 if token.i in end_indices else 0,
+                    )
+                )
+
+        return reference_results
```
I should have caught this earlier, but there's one more major update needed here:

This needs to be modified to handle cases where the predicted and reference tokenization differ. Instead of extracting the reference indices from the reference doc only, you need to align the annotation between the predicted and reference docs and handle cases where the annotation can't be aligned.

Given that you have binary labels on tokens, I think it would look a lot like the loss in the `senter` component. Because you're converting from underlying span annotation, you can't use `Example.get_aligned` directly, so I'm not sure there's a way around a custom alignment method here that does the span+token alignment for this particular case.

You can handle missing values by using an existing loss function from thinc like `SequenceCategoricalCrossentropy`.

thomashacker replied:
I've created a new method `_get_aligned_scores` that validates the alignment between the `predicted` and the `reference` doc and checks for three cases:

1. Tokens are aligned (best case :tada:)
2. Predicted has less tokens than the Reference (`alignment.y2x`)
3. Predicted has more tokens than the Reference (`alignment.x2y`)

It's using the predicted scores and the spans in the reference to create two new lists of scores that are used to calculate the loss. The changes can be found here [18dbe4d](https://github.com/explosion/spacy-experimental/pull/10/commits/18dbe4dd77f38c9fb0d5677e78155046166d1c5d).
There are still some errors when iterating through the alignment properties.

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,352 @@
+                        eg.reference[span.end - 1].idx + len(eg.reference[span.end - 1])
+                    )
+
+            for token in eg.predicted:
+                reference_start_score = 0
+                reference_end_score = 0
+                if token.idx in start_indices:
+                    reference_start_score = 1
+                if token.idx + len(token) in end_indices:
+                    reference_end_score = 1
+
+                reference_results.append(
+                    (
+                        reference_start_score,
+                        reference_end_score,
+                    )
+                )
```
I think you can simplify this to:

```suggestion
            for token in eg.predicted:
                reference_results.append(
                    (
                        1 if token.idx in start_indices else 0,
                        1 if token.idx + len(token) in end_indices else 0,
                    )
                )
```

adrianeboyd commented on spacy_experimental/span_finder/span_finder_component.py:
```diff
@@ -0,0 +1,345 @@
+
+    def _get_aligned_scores(self, examples) -> Tuple[Floats2d, Floats2d]:
+        """Align scores of the predictions to the references for calculating the loss"""
+        # TODO: handle 1-to-many alignments (see xfailed tests)
```
```suggestion
        # TODO: handle misaligned (None) alignments
```

adrianeboyd commented:
Can you rename:

* `candidates_key` -> `predicted_key`
* `reference_key` -> `training_key`

In the project, I'd also like to have `paths.spans_key` as `vars.spans_key` or something else instead, since `paths` seems kind of confusing.

adrianeboyd commented on README.md:
```diff
@@ -216,6 +216,105 @@ annotating_components = ["senter"]
+
+`SpanFinder` uses the following parameters:
+- `threshold`: Probability threshold for predicted spans.
+- `candidates_key`: Name of the [SpanGroup](https://spacy.io/api/spangroup) the predicted spans are saved to.
+- `reference_key`: Name of the [SpanGroup](https://spacy.io/api/spangroup) the reference spans are read from.
```
Rename

adrianeboyd commented on projects/span_finder/README.md:
```diff
@@ -0,0 +1,69 @@
+| `suggester`  | Choose between two suggester architectures (SpanFinder, Ngram)                             |
+| `train`      | Choose a filename for your training data                                                   |
+| `dev`        | Choose a filename for your development data                                                |
+| `span_key`   | Choose a key to specify the SpanGroup for the spancat component to save the predictions to |
```
I think the README needs to be updated?

adrianeboyd commented on projects/span_finder/configs/span_finder/config_tok2vec.cfg:
```diff
@@ -0,0 +1,169 @@
+
+[components.spancat.suggester]
+@misc = "spacy-experimental.span_finder_suggester.v1"
+predicted_key = ${components.span_finder.predicted_key}
```
```suggestion
candidates_key = ${components.span_finder.predicted_key}
```

adrianeboyd commented on projects/span_finder/configs/span_finder/config_trf.cfg:
```diff
@@ -0,0 +1,177 @@
+
+[components.spancat.suggester]
+@misc = "spacy-experimental.span_finder_suggester.v1"
+predicted_key = ${components.span_finder.predicted_key}
```
```suggestion
candidates_key = ${components.span_finder.predicted_key}
```

adrianeboyd replied:
Ah, the setting in `span_finder_suggester` should probably still be called `candidates_key`? Since it's just reading in the candidates, not predicting/saving them to `doc.spans`.

adrianeboyd commented on spacy_experimental/span_finder/tests/test_span_finder.py:
```diff
@@ -0,0 +1,181 @@
+    docs[1].spans[TRAINING_KEY] = [docs[1][3:5]]
+
+    span_finder = nlp.add_pipe(
+        "experimental_span_finder", config={"reference_key": TRAINING_KEY}
```
Rename

adrianeboyd commented on spacy_experimental/span_finder/tests/test_span_finder.py:
```diff
@@ -0,0 +1,181 @@
+        config={
+            "max_length": max_length,
+            "min_length": min_length,
+            "reference_key": TRAINING_KEY,
```
Rename

adrianeboyd commented on spacy_experimental/span_finder/tests/test_span_finder.py:
```diff
@@ -0,0 +1,181 @@
+    docs[0].spans[TRAINING_KEY] = [docs[0][3:4]]
+    docs[1].spans[TRAINING_KEY] = [docs[1][3:5]]
+    span_finder = nlp.add_pipe(
+        "experimental_span_finder", config={"reference_key": TRAINING_KEY}
```
Rename

adrianeboyd commented:
The remaining issues:

* keep `candidates_key` as the key name for the spans suggester
* one more pass through the docs and project for the other key renaming
* a full test of the project with various option combinations after the renaming

thomashacker commented:
Updated the key in the suggester and looked over the docs and the project. I also checked the spaCy project and ran the workflows with different configurations.

adrianeboyd commented on projects/span_finder/README.md:
```diff
@@ -0,0 +1,73 @@
+| `suggester`  | Choose between two suggester architectures (SpanFinder, Ngram)                             |
+| `train`      | Choose a filename for your training data                                                   |
+| `dev`        | Choose a filename for your development data                                                |
+| `span_key`   | Choose a key to specify the SpanGroup for the spancat component to save the predictions to |
```
```suggestion
| `spans_key`   | Choose a key to specify the SpanGroup for the spancat component to save the predictions to |
```

adrianeboyd commented on projects/span_finder/scripts/preprocessing/preprocess_healthsea.py:
```diff
@@ -0,0 +1,82 @@
+    train = []
+    dev = []
+
+    split = int(len(docs) * eval_split)
+    train = docs[split:] + empty_docs[split:]
+    dev = docs[:split] + empty_docs[:split]
```
I don't think this split length works for `empty_docs`?

thomashacker replied:
If it's an empty Doc then `split = 0`

I've tried it this way:

```
nlp = spacy.blank("en")
doc = Doc(nlp.vocab, words=[])

print(doc[:0])
print(doc[0:])
```

And it worked without errors.
